### ISSUE TITLE
Problem Code 128 barcode

### ISSUE THREAD

**Issue by Spelt** [micjahn/ZXing.Net#251]
HI,

Thanks for the great work!

I have a problem with a Code 128 barcode.

If I upload the same file for decode at https://zxing.org/ it decodes fine. Not sure if there is a bug with code 128. I have attached the image with barcode here.

Hope you can help!

Kind Regards,
Edward

![128problem](https://user-images.githubusercontent.com/12660181/77230310-93aa5680-6b93-11ea-8efb-21bfdc0492af.jpg)

**Comment by micjahn** [comment]
There is no guarantee that every barcode can be decoded. And there is no guarantee that the .Net version of zxing gives the same result as the original java version. For example since the beginning of the project I tried to implement a color-to-grayscale-algorithm which works like the java one. But I couldn't remove the last small differences yet. But an identical grayscale image is important for the binarizer to give the same result.

If you shrink the image by 50% the barcode can be decoded successfully.

**Comment by Spelt** [comment]
Thanks! I thought something like this would be the issue.

**Comment by JohnTApp** [comment]
Hi Michael & Edward
My Names John, and I reported the issue. I just saw Micheal's comment here as I was checking the issue. Again, thanks for the great work and for looking at this. I love this tool.
I'm using this in a Delphi FMX application for iOS and Android mobile devices. My Client is a large supermarket and wants to scan shelf labels for inventory counts. (They also use EAN formats for other purposes in the App) All of their shelf labels are code 128. App uses device camera to scan barcode labels on the shelf, so I have no opportunity to scale the image as such. For responsiveness I'm processing about 20fps and Medium Camera resolution which seems to be the sweet spot. None of the shelf code 128 barcodes have successfully scanned at all as yet (whereas EAN is magnificently responsive using ZXING). Is there anything that you can suggest that I could try as a workaround? EG, is there any pre-processing of the image I can perform before parsing to ScanManager (image processing is a bit new to me) or do you think I should perhaps be looking at other means of barcode scanning for my particular use case scenario at this point in time?
Again, Thank you. I do appreciate your time.
Kind Regards,
John

**Comment by micjahn** [comment]
Without any details (like a code snippet of the barcode reader initialization, the selected resolution or a sample image) I can't help here.
The sample image above is taken with a very high resolution. Such images have a lot of noise.
The best opimization in such cases is shrinking the image or reducing the resolution of the camera. For mobile devices I got the best results with a preview resolution of about 1024x1024 pixels.
All other image optimization doesn't make much sense because you get a constantly stream of images of different quality. Normally after a short period of time the barcode reader gets an image in a quality which can be successfully decoded.
The only thing, I can image in your case with Code 128 is, that the barcode reader isn't correctly initialized. Perhaps the possible formats are restricted to EAN and Code 128 is disabled. Doesn't make much sense but it's possible:
var reader = new BarcodeReader { PossibleFormats = new List<BarcodeFormat> { BarcodeFormat.EAN13}};

**Comment by JohnTApp** [comment]
HI Micheal,
Many Thanks.  Below is the code excerpt for everything from using the camera, to initiallizing reader in case I'm doing something wrong. I must admit I thought that the BarcodeFormat was mutually exclusive and that I had to set to BarcodeFormat.Auto if I wanted to check for more that one format. I was filtering out on the return from ReadResult with an ugly IF clause..... so I'll look at that suggestion. I use a TTimer named ScanTimer which fires off every 60ms to get the barcode from a bitmap held in a global bitmap variable. This was the only way that I could find in Delphi that would not constantly throw Access Violations. The code in this prototype needs a little work bit it's quite stable. I'm including here just in case I'm using the ZXING libraries incorrectly, but 
EAN works nicely so I not sure that's the case........ but you may have a better way of initializing. I'll try anything. :-)
With respect to iOS if I set the camera resolution any higher (which you suggest would not be good) the app is very slow. If I set the resolution any lower I don't get a ReadResult from ScanManager (no barcode recognized)..... so I'm snookered on that front.
For Clarity, I'm certainly not asking you to rework the below code. It's just that you mentioned code snippet etc so you can see if my initialization of reader is sufficiently appropriate that I would expect to read a CODE 128. I can write a little app to capture images on the fly at the Resolution I'm setting in the app and send on to you if you would like me to. But if you would rather be left alone as this one may be a bit tricky I totally understand. Please just let me know.
Thanks Again.
John

unit ScanFrmCode;

interface

uses
  System.SysUtils, System.Types, System.UITypes, System.Classes, System.Variants,
  FMX.Types, FMX.Controls, FMX.Forms, FMX.Graphics, FMX.Dialogs, FMX.StdCtrls,
  FMX.ScrollBox, FMX.Memo, FMX.Objects, FMX.Controls.Presentation, FMX.Layouts,
  FMX.MultiView, System.ImageList, FMX.ImgList, FMX.InertialMovement, FMX.Media,
  //System.IOUtils Needed for TPath Etc
  System.IOUtils, FMX.Ani, FMX.WebBrowser, FMX.ExtCtrls,
  // System.Threading Needed for TTask.Run Thread Spawning
  System.Threading,
  // Zing Barcode Libraries
  ZXing.BarcodeFormat,
  ZXing.ReadResult, StrUtils,
  ZXing.ScanManager;

type
  TScanForm = class(TForm)
    ToolBar1: TToolBar;
    ToolBar2: TToolBar;
    MultiView1: TMultiView;
    Button2: TButton;
    SpeedButton1: TSpeedButton;
    SpeedButton2: TSpeedButton;
    SpeedButton3: TSpeedButton;
    ScanBtn: TSpeedButton;
    ToolBar3: TToolBar;
    HomeBtn: TSpeedButton;
    SpeedButton5: TSpeedButton;
    CameraComponent1: TCameraComponent;
    ScanTimer: TTimer;
    SpeedButton4: TSpeedButton;
    Layout1: TLayout;
    a_ImgRectangle: TRectangle;
    VertScrollBox1: TVertScrollBox;
    ImgScaleRectangle: TRectangle;
    imgCameraView: TImage;
    DebugMemo: TMemo;
    ImageList1: TImageList;
    LblImageScanTop: TLabel;
    LblImageScanBottom: TLabel;
    PlotGrid1: TPlotGrid;
    b_ProdutDetailRectangle: TRectangle;
    ImgProduct: TImage;
    LblScanForm: TLabel;
    Label1: TLabel;
    procedure Button1Click(Sender: TObject);
    procedure FormCreate(Sender: TObject);
    procedure Button2Click(Sender: TObject);
    procedure HomeBtnClick(Sender: TObject);
    procedure ScanBtnClick(Sender: TObject);
    procedure CameraComponent1SampleBufferReady(Sender: TObject;
      const ATime: TMediaTime);
    procedure FormDeactivate(Sender: TObject);
    procedure ScanTimerTimer(Sender: TObject);
    procedure FormDestroy(Sender: TObject);
    procedure FormShow(Sender: TObject);
  private
    { Private declarations }
    procedure SetCameraVisibility (VisFlag : Boolean); // Procedure to flip Visibility of Camera Rectangle and children controls on and off
    procedure GetImage;
    Type
      TBarCodeRec = record
        BarcodeNum  : String;
        BarCodeType : TBarCodeFormat;
        BarCodeFound : Boolean;
    end;
    function GetBarcode : TBarCodeRec; // Decodes Barcode and Retuns the Barcode Record from the ReadResult
  public
    { Public declarations }
  end;

var
  ScanForm: TScanForm;

implementation

{$R *.fmx}
Var
  MyBitmap:FMX.Graphics.TBitmap;
  ProductDisplayed : Boolean;

procedure TScanForm.CameraComponent1SampleBufferReady(Sender: TObject;
  const ATime: TMediaTime);
begin
    TThread.Synchronize(TThread.CurrentThread, GetImage);
end;

procedure TScanForm.FormCreate(Sender: TObject);
begin
  MyBitMap:= FMX.Graphics.TBitmap.Create;        // Create The bitmap before calling SetCameraVisibility which uses the created object
end;


procedure TScanForm.FormDestroy(Sender: TObject);
begin
  MyBitmap.Free;
end;

procedure TScanForm.GetImage;
Var
  TMPBitMap : FMX.Graphics.TBitmap;

begin
   try
      TMPBitMap:= FMX.Graphics.TBitmap.Create;
        try
          CameraComponent1.SampleBufferToBitmap(TMPBitMap, True);
          if TMPBitMap.Height < TMPBitMap.Width then TMPBitMap.Rotate(90); // Keep Camera Orientation persistent
          a_ImgRectangle.Fill.Bitmap.Bitmap:=TMPBitMap;   // Assign TMPBitmap to Rectangle for user visibility
          // Put this into the global bitmap for BarCode Scan
          If ScanTimer.Enabled=True then MyBitmap:= a_ImgRectangle.Fill.Bitmap.Bitmap;  // Assign to Global Bitmap for decoding by ScanTimer Thread
        except
          on E : Exception do
            begin
              DebugMemo.Lines.Add('Exception class name = '+E.ClassName);  // Display any exceptions for the prototype in a memo
              DebugMemo.Lines.Add('Exception message = '+E.Message);
              exit;
            end;
        end;
    finally
      TMPBitMap.Free;
    end;
end;

function TScanForm.GetBarCode :TBarCodeRec;
Var
    ReadResult: TReadResult;
    ScanManager: TScanManager;
    ThisBarcode : TBarCodeRec;
begin
  TThread.Synchronize(TThread.CurrentThread,Procedure begin
  ScanManager := TScanManager.Create(TBarcodeFormat.Auto, nil);
  ThisBarcode.BarCodeNum:= '0';
  ThisBarcode.BarCodeType:=Auto;
  ThisBarcode.BarcodeFound:=False;
  try
    try
      ReadResult:=ScanManager.Scan(MyBitMap);
      // Initialise default values into the Result Record for "This Barcode"

      ThisBarcode.BarcodeFound:=False;
      if (ReadResult <> nil) and (ReadResult.text >' ') and ((ReadResult.BarcodeFormat=EAN_13) or (ReadResult.BarcodeFormat=EAN_8)or (ReadResult.BarcodeFormat=UPC_A) or (ReadResult.BarcodeFormat=CODE_128))  then
        begin
           ScanTimer.Enabled:=False;
           ThisBarcode.BarCodeNum:= Readresult.text;
           ThisBarcode.BarCodeType:=ReadResult.BarcodeFormat;
           ThisBarcode.BarcodeFound:=True;
           ProductDisplayed:=True; //  Just Here for Testing. Simulates a product having already been displayed;

        end
      else // ReadResult Is Not Valid
        begin
          ThisBarcode.BarCodeNum:= '0';
          ThisBarcode.BarCodeType:=Auto;
          ThisBarcode.BarcodeFound:=False;

        end;
    except
      on E : Exception do
     { begin
        if E.ClassName='eArgumentException' then    // do nothing
        else    }
         begin
            DebugMemo.Lines.Add('Exception class name = '+E.ClassName); // Display any exception raised and causght in memo for this prototype
            DebugMemo.Lines.Add('Exception message = '+E.Message);
        // end;
         ThisBarcode.BarCodeNum:= '0';
         ThisBarcode.BarCodeType:=Auto;
         ThisBarcode.BarcodeFound:=False;
        //exit;
      end;
    end;
  finally
    begin
       freeandnil(ScanManager);
       Freeandnil(ReadResult);
    end;
  end;
  end);
 Result:=ThisBarcode;
end;

**Comment by micjahn** [comment]
We have a main problem here: you are using a different library: https://github.com/Spelt/ZXing.Delphi/
I don't have any experiences with Delphi. It would be better that you open a new issue there.
If you have a captured image of a code 128 barcode I can only check if it can be decoded with ZXing.Net. Sorry.

**Comment by JohnTApp** [comment]
Oh. I see. I apologize. I saw this and thought you were a Delphi guy. I noticed that Edward had mentioned that he had the same issue in .Net but didn't realize that that was specifically why he had reached out to you. Many Thanks for your correspondence. It's been most appreciated.......... and I also learned something with respect to initializing with the hints. Was just trying to ensure anyone had all the info they needed. I'll leave it with Edward who I believe is the Delphi guru :-) .
Sorry for my misunderstanding.
Kind Regards,
John
